src/shogun/features/DenseFeatures.cpp

```diff
@@ -977,101 +1007,66 @@ template<class ST> bool CDenseFeatures<ST>::is_equal(CDenseFeatures* rhs)
 	return true;
 }
 
-template<class ST> CFeatures* CDenseFeatures<ST>::create_merged_copy(
-		CList* others)
+template <class ST>
+CFeatures* CDenseFeatures<ST>::create_merged_copy(CList* others)
 {
-	SG_DEBUG("entering %s::create_merged_copy()\n", get_name());
+	SG_DEBUG("Entering.\n");
 
-	if (!others)
-		return NULL;
+	REQUIRE(others!=nullptr, "The list of feature instances cannot be null!\n");
 
-	/* first, check other features and count number of elements */
-	CSGObject* other=others->get_first_element();
-	index_t num_vectors_merged=num_vectors;
-	while (other)
+	auto current=others->get_first_element();
+	auto total_num_vectors=get_num_vectors();
+	auto unref_required=others->get_delete_data();
+
+	while (current!=nullptr)
 	{
-		CDenseFeatures<ST>* casted=dynamic_cast<CDenseFeatures<ST>* >(other);
+		auto casted=dynamic_cast<CDenseFeatures<ST>*>(current);
 
-		if (!casted)
-		{
-			SG_ERROR("%s::create_merged_copy(): Could not cast object of %s to "
-					"same type as %s\n",get_name(), other->get_name(), get_name());
-		}
+		REQUIRE(casted!=nullptr, "Could not cast object of %s to same type as %s!\n",
+				current->get_name(), get_name());
+		REQUIRE(num_features==casted->num_features,
+				"Provided feature object has different dimension (%d) than this one (%d)!\n",
+				casted->num_features, num_features);
 
-		if (get_feature_type()!=casted->get_feature_type() ||
-				get_feature_class()!=casted->get_feature_class() ||
-				strcmp(get_name(), casted->get_name()))
-		{
-			SG_ERROR("%s::create_merged_copy(): Features are of different type!\n",
-					get_name());
-		}
+		total_num_vectors+=casted->get_num_vectors();
 
-		if (num_features!=casted->num_features)
-		{
-			SG_ERROR("%s::create_merged_copy(): Provided feature object has "
-					"different dimension than this one\n");
-		}
+		if (unref_required)
+			SG_UNREF(current);
 
-		num_vectors_merged+=casted->get_num_vectors();
-
-		/* check if reference counting is used */
-		if (others->get_delete_data())
-			SG_UNREF(other);
-		other=others->get_next_element();
+		current=others->get_next_element();
 	}
 
-	/* create new feature matrix and copy both instances data into it */
-	SGMatrix<ST> data(num_features, num_vectors_merged);
-
-	/* copy data of this instance */
-	SG_DEBUG("copying matrix of this instance\n")
-	memcpy(data.matrix, feature_matrix.matrix,
-			num_features*num_vectors*sizeof(ST));
+	SGMatrix<ST> data(num_features, total_num_vectors);
+	index_t num_copied=0;
+	copy_feature_matrix(data, num_copied);
+	num_copied+=get_num_vectors();
 
-	/* count number of vectors (not elements) processed so far */
-	index_t num_processed=num_vectors;
+	current=others->get_first_element();
 
-	/* now copy data of other features block wise */
-	other=others->get_first_element();
-	while (other)
+	while (current!=nullptr)
```

Comment:
Yeah iterators would be really good to have. Will have to discuss with you about the scope.